### PR TITLE
General improvements

### PR DESCRIPTION
**Pull request by luisffc**
- Added support for configuring the maximum concurrency limitation, which defaults to 15.
- Removed the column `output` of table `exec_command` in favor of `stdout_output` and `stderr_output` columns.
- Added the columns `stdout_output` and `stderr_output` to table `exec_command` to better reflect the streams of command output.
- Fixed `exec_command` and `exec_command_line` that were always returning as stdout stream even if error, on remote connections

### CODE REVIEW

File: config/exec.spc [hunk 1]
```diff
@@ -3,13 +3,11 @@
 connection "exec_local" {
   plugin = "exec"
 
-  # Working directory to use when running commands.
+  # (local only) Working directory to use when running commands.
```
Comment (cbruno10) [inline config/exec.spc:6]: What happens if a user sets these on a remote connection? Does it error or just do nothing?

Also, in remote connections, what directory do we land in?

File: config/exec.spc [hunk 2]
```diff
@@ -57,6 +55,9 @@ connection "exec_local" {
   # bastion_port = 22
   # bastion_password = "my_password"
   # bastion_private_key = "~/.ssh/my-bastion-host.pem"
+
+  # Optional - Timeout for the connection in seconds, defaults to 15s
+  # timeout = "15s"
```
Comment (cbruno10) [inline config/exec.spc:60]:
```suggestion
  # Optional - Timeout for the connection in seconds, defaults to 15 seconds
  # timeout = "15"
```

To simplify input, we should use an int for this config arg

File: exec/table_exec_command.go [hunk 3]
```diff
@@ -51,30 +54,34 @@ func listExecCommand(ctx context.Context, d *plugin.QueryData, h *plugin.Hydrate
...
+		buf := new(strings.Builder)
+		_, err := io.Copy(buf, outR)
 		if err != nil {
-			plugin.Logger(ctx).Error("listExecCommand", "outputLinesIntoRows", err)
+			plugin.Logger(ctx).Error("listExecCommand", "error reading output", err)
 		}
+		stdout = buf.String()
 	}()
 
+	// wg.Add(1)
```
Comment (cbruno10) [inline exec/table_exec_command.go:77]: Please remove any unused code (if not useful)

File: exec/table_exec_command.go [hunk 4]
```diff
@@ -25,21 +27,22 @@ func tableExecCommand(ctx context.Context) *plugin.Table {
...
 		},
 	}
 }
 
 func listExecCommand(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
-	comm, _, isLocalConnection, err := GetCommunicator(d.Connection)
+	comm, isLocalConnection, err := GetCommunicator(d.Connection)
 	if err != nil {
-		plugin.Logger(ctx).Error("listExecCommand", "command_error", err)
+		plugin.Logger(ctx).Error("listExecCommand", "init", "command_error", err)
```
Comment (cbruno10) [inline exec/table_exec_command.go:41]: Why does this include `init` now?

File: exec/table_exec_command.go [hunk 5]
```diff
@@ -51,30 +54,34 @@ func listExecCommand(ctx context.Context, d *plugin.QueryData, h *plugin.Hydrate
...
 
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		err = outputIntoRow(ctx, d, errR, true)
+		buf := new(strings.Builder)
+		_, err := io.Copy(buf, outR)
 		if err != nil {
-			plugin.Logger(ctx).Error("listExecCommand", "outputLinesIntoRows", err)
+			plugin.Logger(ctx).Error("listExecCommand", "error reading output", err)
```
Comment (cbruno10) [inline exec/table_exec_command.go:72]:
```suggestion
			plugin.Logger(ctx).Error("listExecCommand", "output_read_error", err)
```

File: exec/table_exec_command.go [hunk 6]
```diff
@@ -104,21 +111,21 @@ func listExecCommand(ctx context.Context, d *plugin.QueryData, h *plugin.Hydrate
...
-		Stderr:  errW,
+		// Stderr:  errW,
 	}
 
 	result := commandResult{}
 
 	plugin.Logger(ctx).Debug("listExecCommand", "ctx_done", "cmd.Start...")
 	if err := comm.Start(cmd); err != nil {
-		plugin.Logger(ctx).Error("listExecCommand", "command_error", err)
+		plugin.Logger(ctx).Error("listExecCommand", "comm.Start", "command_error", err)
```
Comment (cbruno10) [inline exec/table_exec_command.go:121]:
```suggestion
		plugin.Logger(ctx).Error("listExecCommand.comm.Start", "command_error", err)
```

File: exec/table_exec_command.go [hunk 7]
```diff
@@ -138,45 +145,83 @@ func listExecCommand(ctx context.Context, d *plugin.QueryData, h *plugin.Hydrate
...
 
-	outputStr := string(output)
-
-	// NOTE - I considered stripping the final newline. The output string looks
-	// weird with the final newline (appears like an extra newline) in SQL
-	// results, so is confusing. But removing the newline is altering the output,
-	// so on balance better to leave it accurate. I'm leaving this code here as a
-	// warning to anyone who is tempted to strip the newline in the future.
-	// outputStr := strings.TrimSuffix(outputStr, "\n")
+	stdoutString := strings.Join(stdoutLines, "\n")
```
Comment (cbruno10) [inline exec/table_exec_command.go:218]: Instead of trimming, we're now using newlines in join, what effect does this have on query results?

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@luisffc Please see comments, thanks!

File: go.mod [hunk 8]
```diff
@@ -1,20 +1,23 @@
 module github.com/turbot/steampipe-plugin-exec
 
-go 1.19
+go 1.21
+
+// replace github.com/turbot/go-exec-communicator => ../../sdks/go-exec-communicator
```
[the remark this answers is not in the document]
Comment (luisffc) [reply go.mod]: removed

File: exec/table_exec_command.go (hunk 7 above)
Comment (luisffc) [reply exec/table_exec_command.go]: fixed

File: config/exec.spc (hunk 1 above)
Comment (luisffc) [reply config/exec.spc:6]: - directory param is ignored on remote connection
- in remote we land in users' home directory

File: exec/table_exec_command.go (hunk 3 above)
Comment (luisffc) [reply exec/table_exec_command.go]: done

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**